src/instructlab/sdg/datamixing.py

```diff
@@ -612,6 +635,11 @@ def collect(
             skills_phase_data = _create_phase10_ds(
                 new_generated_data, self.auxiliary_inst, use_legacy_pretraining_format
             )
+            # XXX(osilkin): To prevent the size of the skills dataset from drowning out the knowledge
+            #               samples, we upsample the knowledge data here to prevent catastrophic forgetting
+            skills_phase_data = upsample_dataset(
+                skills_phase_data, self.upsample_amount
+            )
```

Comment:
Following on from my general comment to point out the specific code here. What if instead of this, we just passed in the `upsample_amount` to the `_gen_leaf_node_data` call a few lines below here as the optional 4th param that's not used today. That sets the sampling size for the dataset, where the default value of `1.0` is unchanged but a value of `0.5` would be sampling 50% of the data, a value of `3.0` would upsample to 300%, and a value of `5000` would upsample to a fixed 5000 samples. I'm not sure the exact right amount of upsampling, but I think the existing code can handle this need.

Reply:
agree with Ben here, in the initial implementation of mixing, we written it such that it would upsample if the provided `--num-samples` or `--sampling-ratio` required it. I guess that got lost somewhere, but that feels like a good way to proceed here.

Reply:
Just doing some quick napkin math, since we found that it performs well when you have 10k resamples out of a 350k dataset, that would put us at a $training/skills$ ratio of roughly 3%. In this case, the rule we should follow would be: 

```py
if len(replay)/len(dataset) < 0.03:
  replay = upsample_ds(replay, len(dataset)*0.03)
```

Reply:
@bbrowning In the current state of the `DataMixer`, it doesn't seem like there's a way for the `_collect` method to know the length of the skills dataset when creating the recipes for the knowledge dataset. If we want to compute the resample amount more dynamically then we would need to refactor this.

Reply:
Yes, refactoring mixing have some knowledge across both skills and knowledge datasets would be a refactor that will be a bit larger lift to get in. Today mixing happens entirely within a single Recipe, and skills and knowledge have different Recipes that are independent of each other.